delphin/util.py

```diff
@@ -272,3 +272,43 @@ def peek(self, n=0, skip=None, drop=False):
             self._buffer_fill(n + 1)
             datum = buffer[n]
         return datum
+
+
+# modified from https://www.python.org/dev/peps/pep-0263/#defining-the-encoding
+_encoding_symbol_re = re.compile(
+    r'^[ \t\f]*.*?coding[:=][ \t]*([-_.a-zA-Z0-9]+)', re.IGNORECASE)
+
+import codecs, io
+
+def detect_encoding(filename, default_encoding='utf-8', comment_char=b';'):
+    encoding = None
+    with io.open(filename, 'rb') as fh:
+        line1 = next(fh)
+        re_match1 = _encoding_symbol_re.search(line1)
+        line1_comment = line1.contains(comment_char)
+        if line1_comment and re_match1:
```

Comment:
`line1.contains()` matches if the comment character is anywhere in the line. E.g., it would probably match the following line (however unlikely):

```
sort := *top*  ; encoding=utf-8
```

Instead, either bake the comment_char into the regex itself (as mentioned above) using `re.escape(comment_char)`, or check something like `line1.lstrip().startswith(comment_char)`